### PR TITLE
cta: docker API

### PR DESCRIPTION
**Pull request by royra**
- passthrough API to docker on port 3001
- websocket impl for exec and logs

### CODE REVIEW

File: packages/compose-tunnel-agent/src/docker-proxy/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
+import http from 'node:http'
+import net from 'node:net'
+import { describe, expect, beforeAll, afterAll, test, jest, it } from '@jest/globals'
+import { ChildProcess, spawn, exec } from 'child_process'
+import pino from 'pino'
+import pinoPretty from 'pino-pretty'
+import Dockerode from 'dockerode'
+import fetch from 'node-fetch'
+import { inspect, promisify } from 'node:util'
+import waitForExpect from 'wait-for-expect'
```
Comment (Yshayy) [inline packages/compose-tunnel-agent/src/docker-proxy/index.test.ts:10]: wait-for-expect seems to be not maintained.
Might worth using alternative or just copying the implementation. (it's a single file)

File: packages/compose-tunnel-agent/src/docker-proxy/index.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    log.warn('invalid upgrade %s', upgrade)
+    socket.end(`Invalid upgrade ${upgrade}`)
+    return undefined
+  })
+
+  return server
+}
+
+// export const createDockerProxy = ({ log: _log, dockerSocket }: {
```
Comment (Yshayy) [inline packages/compose-tunnel-agent/src/docker-proxy/index.ts:75]: Remove comments

File: tunnel-server/src/ssh-server.ts [hunk 3]
```diff
@@ -92,7 +92,7 @@ export const sshServer = (
   const serverEmitter = new EventEmitter() as Omit<SshServer, 'close' | 'listen'>
   const server = new ssh2.Server(
     {
-      debug: x => log.debug(x),
+      // debug: x => log.debug(x),
```
Comment (Yshayy) [inline tunnel-server/src/ssh-server.ts:95]: Is it on purpose?
Remove if not

File: packages/compose-tunnel-agent/src/docker-proxy/ws/handlers/exec.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    const wsStream = createWebSocketStream(ws)
+    wsStream.on('error', err => {
+      const level = err.message.includes('WebSocket is not open') ? 'debug' : 'warn'
+      log[level]('wsStream error %j', inspect(err))
+    })
+
+    if (inspectResults.ProcessConfig.tty) {
+      execStream.pipe(wsStream, { end: false }).pipe(execStream)
+    } else {
+      docker.modem.demuxStream(execStream, wsStream, wsStream)
```
Comment (Yshayy) [inline packages/compose-tunnel-agent/src/docker-proxy/ws/handlers/exec.ts:34]: Do we need to demux the stream if we're writing it to same output?
What's the difference from execStream.pipe(wsStream)

File: packages/compose-tunnel-agent/src/docker-proxy/ws/handlers/logs.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    logStream.on('error', err => {
+      if (err.message !== 'aborted') {
+        log.error('logs stream error %j', inspect(err))
+      }
+    })
+    ws.on('close', () => { abort.abort() })
+
+    const wsStream = createWebSocketStream(ws)
+    wsStream.on('error', err => { log.error('wsStream error %j', inspect(err)) })
+    docker.modem.demuxStream(logStream, wsStream, wsStream)
```
Comment (Yshayy) [inline packages/compose-tunnel-agent/src/docker-proxy/ws/handlers/logs.ts:33]: Same demux question as for the exec. (compared to logStream.pipe(wsStream))

**Review by Yshayy — APPROVED** [review]
Looks good, added few notes

File: packages/compose-tunnel-agent/src/docker-proxy/ws/handlers/exec.ts (hunk 4 above)
Comment (royra) [reply packages/compose-tunnel-agent/src/docker-proxy/ws/handlers/exec.ts:34]: demux is required when `tty=true`, see this [obscure corner of the Docker API doc](https://docs.docker.com/engine/api/v1.41/#tag/Container/operation/ContainerAttach) (under "Stream format"). I'll add a comment.

File: tunnel-server/src/ssh-server.ts (hunk 3 above)
Comment (royra) [reply tunnel-server/src/ssh-server.ts:95]: yes, too much logging

**Pull request merged**